c++/src/capnp/compat/json.c++

```diff
@@ -372,6 +382,251 @@ Orphan<DynamicValue> JsonCodec::decode(
 
 // -----------------------------------------------------------------------------
 
+namespace _ {  // private
+
+class Parser {
+public:
+  Parser(kj::ArrayPtr<const char> input) : input_(input), remaining_(input_) {}
+  void parseValue(JsonValue::Builder &output) {
+    consumeWhitespace();
+    KJ_REQUIRE(remaining_.size() > 0, "JSON message ends prematurely.");
+
+    switch (nextChar()) {
+      case 'n': consume(NULL_); output.setNull();         break;
+      case 'f': consume(FALSE); output.setBoolean(false); break;
+      case 't': consume(TRUE);  output.setBoolean(true);  break;
+      case '"': parseString(output); break;
+      case '[': parseArray(output);  break;
+      case '{': parseObject(output); break;
+      // TODO(security): We could check for numbers more carefully instead of
+      // relying on strtod.
+      default: parseNumber(output); break;
+    }
+  }
+
+  void parseNumber(JsonValue::Builder &output) {
+    // TODO(someday): strtod allows leading +, while JSON grammar does not.
+    // strtod consumes leading whitespace, so we don't have to.
+    char *numEnd;
+    output.setNumber(std::strtod(remaining_.begin(), &numEnd));
```

Comment:
Problems:
- If the first char is not a valid part of a number, then `strtod()` will return zero and set `numEnd` equal to `remaining_.begin()`, i.e. no input will be consumed. I think you'll end up in an infinite loop here. Please add a test.
- If `strtod()` reaches the end of the input, it will keep going past it, since our input buffer is not NUL-terminated.

I think you will need to write a scanner that actually matches the JSON number pattern as one step, then pass it to `strtod()` as a second step.